Code/MinimalLib/minilib.h

```diff
@@ -160,21 +162,62 @@ class JSMol {
       unsigned int maxCutBonds) const;
 #endif
 
-  std::unique_ptr<RDKit::RWMol> d_mol;
   static constexpr int d_defaultWidth = 250;
   static constexpr int d_defaultHeight = 200;
+
+ protected:
+  JSMolBase() = default;
+};
+
+class JSMol : public JSMolBase {
+ public:
+  JSMol() : d_mol(new RDKit::RWMol()) {}
+  JSMol(RDKit::RWMol *mol) : d_mol(mol) { checkNotNull(); }
+  JSMol(const JSMol &other) { d_mol.reset(new RDKit::RWMol(other.get())); }
+  JSMol &operator=(const JSMol &other) {
+    d_mol.reset(new RDKit::RWMol(other.get()));
+    return *this;
+  }
+  RDKit::RWMol &get() const {
+    checkNotNull();
+    return *d_mol.get();
+  }
+
+ private:
+  void checkNotNull() const { CHECK_INVARIANT(d_mol, "d_mol cannot be null"); }
+  std::unique_ptr<RDKit::RWMol> d_mol;
+};
+
+class JSMolShared : public JSMolBase {
+ public:
+  JSMolShared() = delete;
+  JSMolShared(const RDKit::ROMOL_SPTR &mol) : d_mol(mol) { checkNotNull(); }
+  JSMolShared(const JSMolShared &other) { d_mol = other.d_mol; }
+  JSMolShared &operator=(const JSMolShared &other) {
+    d_mol = other.d_mol;
+    return *this;
+  }
+  RDKit::RWMol &get() const {
```

Comment:
I think this should not be `const` since it returns a non-const reference to a data member.

Reply:
Oops, good catch. I was surprised that the compiler did not complain, and then I found that it may happen that the compiler lets it through:
https://isocpp.org/wiki/faq/const-correctness#return-const-ref-from-const-memfn

Reply:
> Oops, good catch. I was surprised that the compiler did not complain, and then I found that it may happen that the compiler lets it through: https://isocpp.org/wiki/faq/const-correctness#return-const-ref-from-const-memfn

I was surprised by that too, but I can never keep that stuff straight

This quote from that FAQ entry is awesome 
> Translation: you need to think. If that scares you, find another line of work; “think” is not a four-letter word.